File: mq/rabbitmq/py/routing/subscriber_saver.py
#!/usr/bin/env python3
#
# Copyright (C) 2019 jayson
#
import json
import logging
import contextlib
import utils.logging
from utils import connection
from config import RABBITMQ_MASTER
from pika.adapters.blocking_connection import BlockingChannel
from pika.spec import Basic, BasicProperties


@contextlib.contextmanager
def consumer_factory(log_file: str):
    with open(log_file, 'a') as w:
        def __consumer(ch: BlockingChannel, method: Basic.Deliver, properties: BasicProperties, body: bytes):
            data = json.loads(body)
            w.write('{:<16}@{:<16}@{:<4}@{:<12}: {}\n'.format(
                data['from'], data['producer'], data['index'], method.routing_key, data['body']))
            ch.basic_ack(delivery_tag=method.delivery_tag)
        yield __consumer


if __name__ == '__main__':
    utils.logging.init_logging(logging.DEBUG)
    with connection.get_exchange_connection(RABBITMQ_MASTER, 'logs_routing', 'direct') as channel:
        channel.basic_qos(prefetch_count=1)
        channel.queue_declare(queue='logs_routing_saver', durable=True, auto_delete=True)
        channel.queue_bind(queue='logs_routing_saver', exchange='logs_routing', routing_key='warning')
        channel.queue_bind(queue='logs_routing_saver', exchange='logs_routing', routing_key='error')

        with consumer_factory('routing.log') as consumer:
            channel.basic_consume('logs_routing_saver', consumer, auto_ack=False)
            channel.start_consuming()

File: mq/rabbitmq/php/src/routing/subscriber_saver.php
<?php
/**
 * rabbitmq hello-world
 */
require_once __DIR__ . '/../bootstrap.php';


$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection(RABBITMQ_MASTER, 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->exchange_declare('logs_routing', 'direct', false, true, false);
$channel->queue_declare('logs_routing_saver', false, true, false, true);
$channel->queue_bind('logs_routing_saver', 'logs_routing', 'warning');
$channel->queue_bind('logs_routing_saver', 'logs_routing', 'error');

$channel->basic_qos(null, 1, true);
$channel->basic_consume('logs_routing_saver', null, false, false, false, false,
    function(\PhpAmqpLib\Message\AMQPMessage $message) use ($channel) {
        $data = json_decode($message->body, true);
        file_put_contents('routing.log', sprintf("%-12s %-12s %-8s %-12s: %s\n", $data['from'], $data['producer'], $data['index'],
            $message->delivery_info['routing_key'], $data['body']), FILE_APPEND);
        $channel->basic_ack($message->getDeliveryTag());
    }
);

while ($channel->is_consuming()) {
    $channel->wait();
}
